Mathlib/Data/Setoid/Basic.lean

```diff
@@ -0,0 +1,487 @@
+
+/-- Given a map `f : N → M` and an equivalence relation `r` on `β`, the equivalence relation
+    induced on `α` by `f` equals the kernel of `r`'s quotient map composed with `f`. -/
+theorem comap_eq {f : α → β} {r : Setoid β} : comap f r = ker (@Quotient.mk'' _ r ∘ f) :=
+  ext fun x y => show _ ↔ ⟦_⟧ = ⟦_⟧ by rw [Quotient.eq]; rfl
+#align setoid.comap_eq Setoid.comap_eq
+
+/-- The second isomorphism theorem for sets. -/
+noncomputable def comapQuotientEquiv (f : α → β) (r : Setoid β) :
+    Quotient (comap f r) ≃ Set.range (@Quotient.mk'' _ r ∘ f) :=
+  (Quotient.congrRight <| ext_iff.1 comap_eq).trans <| quotientKerEquivRange <| Quotient.mk'' ∘ f
+#align setoid.comap_quotient_equiv Setoid.comapQuotientEquiv
+
+variable (r f)
+
+/-- The third isomorphism theorem for sets. -/
+def quotientQuotientEquivQuotient (s : Setoid α) (h : r ≤ s) :
+    Quotient (ker (Quot.mapRight h)) ≃ Quotient s where
+  toFun x :=
+    (Quotient.liftOn' x fun w =>
+        (Quotient.liftOn' w (@Quotient.mk'' _ s)) fun x y H => Quotient.sound <| h H)
+      fun x y => Quotient.inductionOn₂' x y fun w z H => show @Quot.mk _ _ _ = @Quot.mk _ _ _ from H
+  invFun x :=
+    (Quotient.liftOn' x fun w => @Quotient.mk'' _ (ker <| Quot.mapRight h) <| @Quotient.mk'' _ r w)
+      fun x y H => Quotient.sound' <| show @Quot.mk _ _ _ = @Quot.mk _ _ _ from Quotient.sound H
+  left_inv x :=
+    Quotient.inductionOn' x fun y => Quotient.inductionOn' y fun w => by show ⟦_⟧ = _; rfl
+  right_inv x := Quotient.inductionOn' x fun y => by show ⟦_⟧ = _; rfl
+#align setoid.quotient_quotient_equiv_quotient Setoid.quotientQuotientEquivQuotient
+
+variable {r f}
+
+open Quotient
+
+/-- Given an equivalence relation `r` on `α`, the order-preserving bijection between the set of
+equivalence relations containing `r` and the equivalence relations on the quotient of `α` by `r`. -/
+def correspondence (r : Setoid α) : { s // r ≤ s } ≃o Setoid (Quotient r) where
+  toFun s := mapOfSurjective s.1 Quotient.mk'' ((ker_mk_eq r).symm ▸ s.2) exists_rep
+  invFun s := ⟨comap Quotient.mk' s, fun x y h => by rw [comap_rel, eq_rel.2 h]⟩
+  left_inv s :=
+    Subtype.ext_iff_val.2 <|
+      ext' fun _ _ =>
+        ⟨fun h =>
+          let ⟨a, b, hx, hy, H⟩ := h
+          s.1.trans' (s.1.symm' <| s.2 <| eq_rel.1 hx) <| s.1.trans' H <| s.2 <| eq_rel.1 hy,
+          fun h => ⟨_, _, rfl, rfl, h⟩⟩
+  right_inv s :=
+    let Hm : ker Quotient.mk' ≤ comap Quotient.mk' s := fun x y h => by
+      rw [comap_rel, (@eq_rel _ r x y).2 (ker_mk_eq r ▸ h)]
+    ext' fun x y =>
+      ⟨fun h => let ⟨a, b, hx, hy, H⟩ := h; hx ▸ hy ▸ H,
+        Quotient.inductionOn₂ x y fun w z h => ⟨w, z, rfl, rfl, h⟩⟩
+  map_rel_iff' := -- fun s t =>
+    ⟨fun h x y hs =>
+      let ⟨a, b, hx, hy, ht⟩ := h ⟨x, y, rfl, rfl, hs⟩
+      t.1.trans' (t.1.symm' <| t.2 <| eq_rel.1 hx) <| t.1.trans' ht <| t.2 <| eq_rel.1 hy,
+      fun h x y hs =>
+      let ⟨a, b, hx, hy, Hs⟩ := hs
+      ⟨a, b, hx, hy, h Hs⟩⟩
```

Comment:
This is the last sorry in the file, and I'm a bit stumped here -- if anyone wants to have a crack at it, go for it!